context.go

```diff
@@ -0,0 +1,111 @@
+package turn
+
+import (
+	"crypto/md5" // #nosec
+	"net"
+	"strings"
+
+	"github.com/pion/stun"
+	"github.com/pkg/errors"
+)
+
+// Context is used to hold the context of a request
+type Context struct {
+	srcAddr net.Addr
+	dstAddr net.Addr
+	msg     *stun.Message
+	conn    net.PacketConn
+
+	realm       string
+	authHandler AuthHandler
+
+	messageIntegrity stun.MessageIntegrity
+}
+
+// NewContext creates a new instance of Context.
+func NewContext(conn net.PacketConn, srcAddr net.Addr, m *stun.Message, realm string, authHandler AuthHandler) *Context {
+	return &Context{
+		conn:        conn,
+		srcAddr:     srcAddr,
+		dstAddr:     conn.LocalAddr(),
+		msg:         m,
+		realm:       realm,
+		authHandler: authHandler,
+	}
+}
+
+// Authenticate by AttrMessageIntegrity and return error
+func (ctx *Context) Authenticate() error {
+	if !ctx.msg.Contains(stun.AttrMessageIntegrity) {
+		nonce, err := buildNonce()
+		if err != nil {
+			return err
+		}
+
+		return ctx.Return(
+			stun.ClassErrorResponse,
+			&stun.ErrorCodeAttribute{Code: stun.CodeUnauthorized},
+			stun.NewNonce(nonce),
+			stun.NewRealm(ctx.realm))
+	}
+
+	var ourKey [16]byte
+	nonceAttr := &stun.Nonce{}
+	usernameAttr := &stun.Username{}
+	realmAttr := &stun.Realm{}
+
+	if err := realmAttr.GetFrom(ctx.msg); err != nil {
+		return ctx.ResponseWithError(err, stun.CodeBadRequest)
+	}
+
+	if err := nonceAttr.GetFrom(ctx.msg); err != nil {
+		return ctx.ResponseWithError(err, stun.CodeBadRequest)
+	}
+
+	if err := usernameAttr.GetFrom(ctx.msg); err != nil {
+		return ctx.ResponseWithError(err, stun.CodeBadRequest)
+	}
+
+	password, ok := ctx.authHandler(usernameAttr.String(), ctx.srcAddr)
+	if !ok {
+		return ctx.ResponseWithError(errors.Errorf("No user exists for %s", usernameAttr.String()), stun.CodeBadRequest)
+	}
+
+	/* #nosec */
+	ourKey = md5.Sum([]byte(usernameAttr.String() + ":" + realmAttr.String() + ":" + password))
+	if err := assertMessageIntegrity(ctx.msg, ourKey[:]); err != nil {
+		return ctx.ResponseWithError(err, stun.CodeBadRequest)
+	}
+
+	ctx.messageIntegrity = stun.NewLongTermIntegrity(usernameAttr.String(), realmAttr.String(), password)
+
+	return nil
+}
+
+// Response is a help method to send success response.
+func (ctx *Context) Response(attrs ...stun.Setter) error {
+	return ctx.Return(stun.ClassSuccessResponse, attrs...)
+}
+
+// ResponseWithError is a help method to send error response with error.
+func (ctx *Context) ResponseWithError(err error, errorCode stun.ErrorCode, attrs ...stun.Setter) error {
+	attrs = append(attrs, &stun.ErrorCodeAttribute{Code: errorCode})
+	if sendErr := ctx.Return(stun.ClassErrorResponse,
+		attrs...,
+	); sendErr != nil {
+		err = errors.Errorf(strings.Join([]string{sendErr.Error(), err.Error()}, "\n"))
+	}
+
+	return err
+}
+
+// Return is a help method to send response message.
+func (ctx *Context) Return(class stun.MessageClass, attrs ...stun.Setter) error {
```

Comment:
Would this method need to be public?
The name 'return' does not clearly represent what it does...  can we rename it as 'buildAndSend'?

Reply:
rename to buildAndSend